README.rst
.. image:: https://img.shields.io/pypi/wheel/logfury.svg
    :target: https://pypi.python.org/pypi/logfury/
.. image:: https://img.shields.io/pypi/l/logfury.svg
    :target: https://pypi.python.org/pypi/logfury/
.. image:: https://img.shields.io/pypi/v/logfury.svg
    :target: https://pypi.python.org/pypi/logfury/
.. image:: https://img.shields.io/pypi/dm/logfury.svg
    :target: https://pypi.python.org/pypi/logfury/
.. image:: https://github.com/reef-technologies/logfury/workflows/CI/badge.svg?branch=master
    :target: https://github.com/reef-technologies/logfury/actions/workflows/ci.yml

========
Logfury
========

Logfury is a tool for python library maintainers. It allows for responsible, low-boilerplate logging of method calls.

*****************************
whats with the weird import
*****************************

.. sourcecode:: python

    from logfury.v1 import DefaultTraceMeta

If you were to use logfury in your library, any change to the API could potentially break your program. Nobody wants that.

Thanks to this import trick I can keep the API very stable. At the same time I can change the functionality of the library and change default behavior of next middle version, without changing the name of the package. This way YOU decide when to adopt potentially incompatible API changes, by incrementing the API version on import. This concept is called "apiver".


*****************
Installation
*****************

^^^^^^^^^^^^^^^^^^^^
Current stable
^^^^^^^^^^^^^^^^^^^^

::

    pip install logfury

^^^^^^^^^^^^^^^^^^^^
Development version
^^^^^^^^^^^^^^^^^^^^

::

    git clone git@github.com:reef-technologies/logfury.git
    pip install -e .


*****************
Basic usage
*****************

^^^^^^^^^^^^^^^^^^^^^^^^^^^
DefaultTraceMeta metaclass
^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. sourcecode:: pycon

    >>> import logging
    >>> import six
    >>>
    >>> from logfury.v1 import DefaultTraceMeta, limit_trace_arguments, disable_trace
    >>>
    >>>
    >>> logging.basicConfig()
    >>> logger = logging.getLogger(__name__)
    >>> logger.setLevel(logging.DEBUG)
    >>>
    >>>
    >>> @six.add_metaclass(DefaultTraceMeta)
    >>> class Foo:
    ...     def baz(self, a, b, c=None):
    ...         return True
    ...     def get_blah(self):
    ...         return 5
    ...     def _hello(self):
    ...         pass
    ...     @disable_trace
    ...     def world(self):
    ...         pass
    ...     def __repr__(self):
    ...         return '<{} object>'.format(self.__class__.__name__,)
    ...
    >>> class Bar(Foo):
    ...     def baz(self, a, b, c=None):
    ...         b += 1
    ...         return super(Bar, self).baz(a, b, c)
    ...     def world(self):
    ...         pass
    ...     @limit_trace_arguments(skip=['password'])
    ...     def secret(self, password, other):
    ...         pass
    ...     @limit_trace_arguments(only=['other'])
    ...     def secret2(self, password, other):
    ...         pass
    ...
    >>> a = Foo()
    >>> a.baz(1, 2, 3)
    DEBUG:__main__:calling Foo.baz(self=<Foo object>, a=1, b=2, c=3)
    >>> a.baz(4, b=8)
    DEBUG:__main__:calling Foo.baz(self=<Foo object>, a=4, b=8)
    >>> a.get_blah()  # nothing happens, since v1.DefaultTraceMeta does not trace "get_.*"
    >>> a._hello()  # nothing happens, since v1.DefaultTraceMeta does not trace "_.*"
    >>> a.world()  # nothing happens, since v1.DefaultTraceMeta does not trace "_.*"
    >>> b = Bar()
    >>> b.baz(4, b=8)  # tracing is inherited
    DEBUG:__main__:calling Bar.baz(self=<Bar object>, a=4, b=8)
    DEBUG:__main__:calling Foo.baz(self=<Bar object>, a=4, b=9, c=None)
    >>> b.world()  # nothing happens, since Foo.world() tracing was disabled and Bar inherited it
    >>> b.secret('correct horse battery staple', 'Hello world!')
    DEBUG:__main__:calling Bar.secret(self=<Bar object>, other='Hello world!') (hidden args: password)
    >>> b.secret2('correct horse battery staple', 'Hello world!')
    DEBUG:__main__:calling Bar.secret2(other='Hello world!') (hidden args: self, password)


^^^^^^^^^^^^^^^^^^^^
trace_call decorator
^^^^^^^^^^^^^^^^^^^^

.. sourcecode:: pycon

    >>> import logging
    >>> from logfury import *
    >>> logging.basicConfig()
    >>> logger = logging.getLogger(__name__)
    >>>
    >>> @trace_call(logger)
    ... def foo(a, b, c=None):
    ...     return True
    ...
    >>> foo(1, 2, 3)
    True
    >>> logger.setLevel(logging.DEBUG)
    >>> foo(1, 2, 3)
    DEBUG:__main__:calling foo(a=1, b=2, c=3)
    True
    >>> foo(1, b=2)
    DEBUG:__main__:calling foo(a=1, b=2)
    True
    >>>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2021-10-23

### Fixed
* Fixed decorating classes
* Fixed arguments ordering in logs

## [1.0.0] - 2021-10-19

### Added
* ApiVer `v1`
* Support for Python >3.6 up to 3.10
* Nox task manager
* GitHub Actions CI
* GitHub Actions CD

### Fixed
* Handling `staticmethod` and `classmethod`
* Support for decorating `class`
* Logging traced object names

### Deprecated
* Python <3.5 support

[Unreleased]: https://github.com/Backblaze/b2-sdk-python/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/Backblaze/b2-sdk-python/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/Backblaze/b2-sdk-python/compare/0.1.2...v1.0.0


noxfile.py
import os
from glob import glob

import nox

CI = os.environ.get('CI') is not None
NOX_PYTHONS = os.environ.get('NOX_PYTHONS')

PYTHON_VERSIONS = [
    '3.5',
    '3.6',
    '3.7',
    '3.8',
    '3.9',
    '3.10',
] if NOX_PYTHONS is None else NOX_PYTHONS.split(',')
PYTHON_DEFAULT_VERSION = PYTHON_VERSIONS[-1]

PY_PATHS = ['logfury', 'test', 'noxfile.py', 'setup.py']

REQUIREMENTS_FORMAT = ['yapf==0.27']
REQUIREMENTS_LINT = [*REQUIREMENTS_FORMAT, 'flake8==4.0.1', 'pytest==6.2.5']
REQUIREMENTS_TEST = [
    "pytest==6.2.5;python_version>'3.5'",
    "pytest==6.1.1;python_version=='3.5'",
    "pytest-cov==3.0.0;python_version>'3.5'",
    "pytest-cov==2.10.1;python_version=='3.5'",
    "testfixtures==6.18.3",
]
REQUIREMENTS_COVER = ['cover']
REQUIREMENTS_BUILD = ['setuptools>=20.2']

nox.options.reuse_existing_virtualenvs = True
nox.options.sessions = [
    'lint',
    'test',
]

# In CI, use Python interpreter provided by GitHub Actions
if CI:
    nox.options.force_venv_backend = 'none'


def install_myself(session, extras=None):
    """Install from the source."""
    arg = '.'
    if extras:
        arg += '[{}]'.format(','.join(extras))

    session.install('-e', arg)


@nox.session(name='format', python=PYTHON_DEFAULT_VERSION)
def format_(session):
    """Format the code."""
    session.install(*REQUIREMENTS_FORMAT)
    session.run('yapf', '--in-place', '--parallel', '--recursive', *PY_PATHS)


@nox.session(python=PYTHON_DEFAULT_VERSION)
def lint(session):
    """Run linters."""
    install_myself(session)
    session.install(*REQUIREMENTS_LINT)
    session.run('yapf', '--diff', '--parallel', '--recursive', *PY_PATHS)
    session.run('flake8', *PY_PATHS)


@nox.session(python=PYTHON_VERSIONS)
def test(session):
    """Run unit tests."""
    install_myself(session)
    session.install(*REQUIREMENTS_TEST)
    args = ['--cov=logfury', '--cov-branch', '--cov-report=xml', '--doctest-modules']
    session.run('pytest', *args, *session.posargs, 'test')

    if not session.posargs:
        session.notify('cover')


@nox.session
def cover(session):
    """Perform coverage analysis."""
    session.install(*REQUIREMENTS_COVER)
    session.run('coverage', 'report', '--fail-under=75', '--show-missing', '--skip-covered')
    session.run('coverage', 'erase')


@nox.session(python=PYTHON_DEFAULT_VERSION)
def build(session):
    """Build the distribution."""
    session.install(*REQUIREMENTS_BUILD)
    session.run('python', 'setup.py', 'check', '--metadata', '--strict')
    session.run('rm', '-rf', 'build', 'dist', 'logfury.egg-info', external=True)
    session.run('python', 'setup.py', 'bdist_wheel', *session.posargs)

    # Set outputs for GitHub Actions
    if CI:
        asset_path = glob('dist/*')[0]
        print('::set-output name=asset_path::', asset_path, sep='')

        version = os.environ['GITHUB_REF'].replace('refs/tags/v', '')
        print('::set-output name=version::', version, sep='')
